### PR TITLE
Adds basic functionality of ToDo list

### PR DESCRIPTION
**Pull request by iranda**
**Description**

- Fixed bugs;
- Added features: one/all task completion, one/all task deletion;
- Added basic unit tests for: Todo class, server;
- Added styles (including mobile devices);
- Implemented basic unit tests;

**Todo**
- Add tests for todoCtrl;
- Expand cache options: save user actions offline and send diff to save on server. Problem: organize correct merge with all clients updates.
- Host solution.

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -1,32 +1,164 @@
 const server = io('http://localhost:3003/');
 const list = document.getElementById('todo-list');
 
+var AdditionOrder = Object.freeze({ APPEND: 0, PREPEND: 1});
```
Comment (davidhariri) [inline client.js]: I didn't even know JavaScript had this! Nice one! 🎉

File: client.js [hunk 2]
```diff
@@ -1,32 +1,164 @@
...
+   };
+};
+
 // NOTE: These are all our globally scoped functions for interacting with the server
 // This function adds a new todo from the input
 function add() {
-    console.warn(event);
     const input = document.getElementById('todo-input');
 
+    if (input.value == "") return;
```
Comment (davidhariri) [inline client.js]: Nice detail 👍

File: client.js [hunk 3]
```diff
@@ -1,32 +1,164 @@
 const server = io('http://localhost:3003/');
 const list = document.getElementById('todo-list');
 
+var AdditionOrder = Object.freeze({ APPEND: 0, PREPEND: 1});
+var isFirstConnection = true;
+
+document.body.onload = function() {
+   var input = document.getElementById('todo-input');
+   input.onkeyup = function(e) {
+    if(e.keyCode == 13) // Enter key
```
Comment (davidhariri) [inline client.js]: Nice detail 👍

File: client.js [hunk 4]
```diff
@@ -1,32 +1,164 @@
...
 
-function render(todo) {
-    console.log(todo);
+// This function removes selected todo item
+function remove() {
+    var listItem = this.parentElement;
+    var id = listItem.id;
+
+    // Emit todo id to server
+    server.emit('removeTodo', id);
```
Comment (davidhariri) [inline client.js]: ~Why not just use `this` like you have in `check()`?~
Ah, I get it. Nice. Semantic, but could just put `listItem.id` in here directly

File: client.js [hunk 5]
```diff
@@ -1,32 +1,164 @@
...
+function removeAll() {
+    server.emit('removeAll');
+}
+
+// This function checks selected todo item
+function check() {
+    server.emit('checkTodo', this.id);
+}
+
+// This function checks all elements in the list
```
Comment (davidhariri) [inline client.js]: Thanks for commenting 👍

File: client.js [hunk 6]
```diff
@@ -1,32 +1,164 @@
...
+function checkAll() {
+    server.emit('checkAll');
+}
+
+// This function adds new element(li) to the list depending of the order(AdditionOrder)
+function render(todo, ao = AdditionOrder.APPEND) {
     const listItem = document.createElement('li');
+    listItem.id = todo._id;
+    listItem.className = (todo.isChecked) ? 'todo-checked' : 'todo-unchecked';
```
Comment (davidhariri) [inline client.js]: Good place to use a ternary 👍

File: client.js [hunk 7]
```diff
@@ -1,32 +1,164 @@
...
     listItem.appendChild(listItemText);
-    list.append(listItem);
+
+    var removeSpan = document.createElement('span');
+    removeSpan.className = 'todo-remove';
+    removeSpan.appendChild( document.createTextNode("\u00D7") );
+    removeSpan.addEventListener('click', remove);
+    listItem.append(removeSpan);
+
+    if (ao == AdditionOrder.APPEND) list.append(listItem);
```
Comment (davidhariri) [inline client.js]: Love how you made the AdditionOrder an enum option. Nice detail 👍

File: client.js [hunk 8]
```diff
@@ -1,32 +1,164 @@
...
+// Check all items
+var cachCheckAll = () => {
+    var cachedItems = cachGet().forEach((todo) => { cachCheckItem(todo._id) } );
+};
+
+//Get cached data
+var cachGet = () => {
+    var values = [], data = [];
+    var keys = Object.keys(localStorage).map((key) => parseInt(key));
+    keys.sort(function(a,b){return b - a}); // to enshure that the orer of items is kept
```
Comment (davidhariri) [inline client.js]: Nice detail 👍

**Comment by NicHaley** [comment]
Really impressed by your code! I had a look at the Client-side stuff. You completed all the tasks and even added in some great new features (caching the todos in local storage). My only suggestion would have been to use a CSS preprocessor like SASS. Also being able to toggle off all todos would have been nice too. Really great work though!

File: client.js (hunk 1 above)
Comment (iranda) [inline client.js]: I'm happy that my work helped you to learn something new 😊

File: client.js (hunk 4 above)
Comment (iranda) [inline client.js]: Of course `listItem.id` would be better.

File: client.js (hunk 5 above)
Comment (iranda) [inline client.js]: I tried to kept your style in commenting 😊

File: client.js (hunk 7 above)
Comment (iranda) [inline client.js]: Yep, I like enums😏
They could be very helpful for easy-understanding.

File: client.js (hunk 3 above)
Comment (iranda) [inline client.js]: I think that button for addition new items is redundant and would be more user-friendly just to handle 'enter'.

File: client.js (hunk 8 above)
Comment (iranda) [inline client.js]: It could be unexpectedly if user got ToDo items not in chronological order

**Comment by iranda** [comment]
Hey David, 
Thank you for detailed review of my solution. I'll take a look on the defects.
Sorry for burning DB in tests. I know about this tricky thing, but my hands didn't reach up to it. My fault.

Hey Nic,
Thank you for taking a look on the code. I will look in the SASS direction after using React.
I also thought about unchecking all todos, so I need to finish this feature.

**Pull request closed without merging**